### Задача 1

1. Скачайте данные по следующей [ссылке](https://www.kaggle.com/datasets/ionaskel/laptop-prices).
2. Считайте данные с помощью библиотеки pandas.
3. Выведите на экран первые 5 строк.

In [3]:
import pandas as pd

# Считываем данные
data = pd.read_csv(r'i:\!GeekBrains\Business Intelligence\II четверть Специализация\14 Python для аналитиков\Урок02Семинар\kc-house-data.csv', encoding='ISO-8859-1')

# Выводим первые 5 строк
data.head()


id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### Задача 1.2

Создайте новый признак `delta_renovated`, который будет содержать разницу в годах между годом реновации дома (`yr_renovated`) и годом постройки дома (`yr_built`). Если реновации дома не было (год реновации равен 0), то в новом признаке должно быть значение 0.

In [4]:
# Создаем признак delta_renovated
data['delta_renovated'] = data['yr_renovated'] - data['yr_built']
data.loc[data['yr_renovated'] == 0, 'delta_renovated'] = 0

# Выводим первые 5 строк для проверки
data.head()


id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  delta_renovated  
0        5650                0  
1        7639               40  
2        8062                0  
3        5000                0  
4        7503                0  

[5 rows x 22 columns]

### Задача 1.3

Создайте два новых признака: `year_sold` и `month_sold`, которые будут содержать год и месяц продажи соответственно, извлеченные из признака `date`.


In [5]:
# Преобразуем столбец date в формат datetime
data['date'] = pd.to_datetime(data['date'])

# Создаем признак year_sold
data['year_sold'] = data['date'].dt.year

# Создаем признак month_sold
data['month_sold'] = data['date'].dt.month

# Выводим первые 5 строк для проверки
data[['date', 'year_sold', 'month_sold']].head()


date  year_sold  month_sold
0 2014-10-13       2014          10
1 2014-12-09       2014          12
2 2015-02-25       2015           2
3 2014-12-09       2014          12
4 2015-02-18       2015           2

### Задача 1.4

Удалите из датафрейма признаки: `date`, `zipcode`, `lat`, `long`.


In [6]:
# Удаляем указанные признаки
data.drop(['date', 'zipcode', 'lat', 'long'], axis=1, inplace=True)

# Выводим первые 5 строк для проверки
data.head()


id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0           0     0          3      7        1180              0      1955   
1           0     0          3      7        2170            400      1951   
2           0     0          3      6         770              0      1933   
3           0     0          5      7        1050            910      1965   
4           0     0          3      8        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  delta_renovated  year_sold  \
0             0           1340        5650                0       2014   
1          1991           1690        7639               40       2014   
2             0           2720        8062                0       2015   
3             0           1360        5000                0       2014   
4             0           1800        7503                0       2015   

   month_sold  
0          10  
1          12  
2           2  
3          12  
4           2

### Задача 2

Создайте датафрейм с клиентами:

In [7]:
clients = pd.DataFrame({
    'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
                  939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
                  2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
    'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
                 1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
                 5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
                 5083000375, 7920100025, 1951600150, 809001400, 339600110,
                 1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
                 8861700030, 3303980210, 7731100066, 8146100580, 825069097,
                 3889100029, 9524100196]
})


### Задача 2.1

Присоедините к таблице `clients` данные по домам через метод `join`.

In [11]:
# Устанавливаем id в качестве индекса для таблицы data (похоже, что у вас нет колонки 'house_id', а есть 'id')
data.set_index('id', inplace=True)

# Устанавливаем house_id в качестве индекса для таблицы clients
clients.set_index('house_id', inplace=True)

# Присоединяем таблицы
result = clients.join(data)

# Возвращаем индекс в датафреймы
clients.reset_index(inplace=True)
data.reset_index(inplace=True)

result.head()  # Отображаем первые строки полученного датафрейма


client_id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
house_id                                                                      
339600110        4809  395000.0         3       2.50         1610      3755   
809001400        3113  925000.0         3       1.00         1630      3200   
825069097        1764  770000.0         3       2.50         2650     40705   
1099600250       2242  260000.0         3       1.75         1710      6400   
1189000207       2308  387000.0         2       2.50         1170      1394   

            floors  waterfront  view  condition  grade  sqft_above  \
house_id                                                             
339600110      2.0           0     0          3      7        1610   
809001400      1.5           0     0          4      8        1630   
825069097      2.0           0     0          3      9        2650   
1099600250     1.0           0     0          4      7        1240   
1189000207     2.0           0     0          3      8        1170   

            sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
house_id                                                                       
339600110               0      1987             0           1300        3823   
809001400               0      1912             0           1710        3600   
825069097               0      1994             0           2550       42625   
1099600250            470      1976             0           1600        6400   
1189000207              0      2001             0           1250        3136   

            delta_renovated  year_sold  month_sold  
house_id                                            
339600110                 0       2014           9  
809001400                 0       2014           9  
825069097                 0       2014           6  
1099600250                0       2014          12  
1189000207                0       2014          10

### Задача 2.2

Присоедините к таблице `clients` данные по домам через метод `merge`.

In [14]:
merged_data = clients.merge(data, how='left', left_on='house_id', right_on='id')
merged_data.head()


house_id  client_id          id     price  bedrooms  bathrooms  \
0  8965450190       1459  8965450190  295000.0         3       2.50   
1  6823100225       4684  6823100225  700000.0         4       1.75   
2  5104540330       3498  5104540330  679000.0         4       2.50   
3  2131701075       3942  2131701075  420000.0         3       1.75   
4  1522700060       4535  1522700060  518000.0         4       2.75   

   sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0         1500      3060     2.0           0  ...      7        1500   
1         1870      6000     1.0           0  ...      8        1670   
2         3680      7236     2.0           0  ...     10        3680   
3         1720      5000     1.5           0  ...      8        1720   
4         2520     14021     2.0           0  ...      9        2520   

   sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
0              0      1994             0           2700        7734   
1            200      1949             0           1710        6000   
2              0      2006             0           3310        7180   
3              0      1932          2009           1410        8300   
4              0      1999             0           2330       14007   

   delta_renovated  year_sold  month_sold  
0                0       2015           2  
1                0       2015           4  
2                0       2015           5  
3               77       2014          12  
4                0       2014           6  

[5 rows x 22 columns]

### Задача 3

Задача 3.1
Найдите среднюю стоимость домов в зависимости от количества спален. Отсортируйте от меньшей стоимости к большей.

In [17]:
data.groupby('bedrooms')['price'].mean().sort_values()


bedrooms
1     3.176429e+05
2     4.013727e+05
0     4.095038e+05
3     4.662321e+05
11    5.200000e+05
4     6.354195e+05
33    6.400000e+05
5     7.865998e+05
10    8.193333e+05
6     8.255206e+05
9     8.939998e+05
7     9.511847e+05
8     1.105077e+06
Name: price, dtype: float64

Задача 3.2
Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома.

In [18]:
data.groupby('condition')['price'].agg(['min', 'mean', 'max'])


min           mean        max
condition                                    
1           78000.0  334431.666667  1500000.0
2           80000.0  327287.145349  2555000.0
3           75000.0  542012.578148  7062500.0
4           89000.0  521200.390033  7700000.0
5          110000.0  612418.089359  3650000.0

Задача 3.3
Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида.

In [20]:
data.groupby(['waterfront', 'view'])['id'].count().unstack()


view              0      1      2      3      4
waterfront                                     
0           19489.0  331.0  955.0  491.0  184.0
1               NaN    1.0    8.0   19.0  135.0

Задача 3.4
Каких домов в зависимости от этажности и количества спален больше?

In [21]:
data.groupby(['floors', 'bedrooms'])['id'].count().unstack().fillna(0)


bedrooms   0      1       2       3       4      5      6     7    8    9   \
floors                                                                       
1.0       4.0  162.0  1951.0  5455.0  2383.0  605.0  104.0   9.0  5.0  0.0   
1.5       0.0   21.0   182.0   786.0   698.0  185.0   30.0   7.0  1.0  0.0   
2.0       6.0   12.0   497.0  3118.0  3682.0  775.0  119.0  19.0  6.0  4.0   
2.5       0.0    1.0     5.0    56.0    58.0   23.0   14.0   2.0  0.0  2.0   
3.0       2.0    3.0   123.0   405.0    61.0   13.0    5.0   1.0  0.0  0.0   
3.5       1.0    0.0     2.0     4.0     0.0    0.0    0.0   0.0  1.0  0.0   

bedrooms   10   11   33  
floors                   
1.0       1.0  0.0  1.0  
1.5       0.0  0.0  0.0  
2.0       2.0  1.0  0.0  
2.5       0.0  0.0  0.0  
3.0       0.0  0.0  0.0  
3.5       0.0  0.0  0.0

Из этой таблицы можно сделать следующие выводы:

1. Большинство одноэтажных домов имеют 3 спальни (5455 домов).
2. Дома с 1.5 этажами чаще всего также имеют 3 спальни (786 домов).
3. Для двухэтажных домов наиболее типично наличие 4 спален (3682 дома).
4. Дома с 2.5 и 3 этажами также чаще всего имеют 3 спальни.

Интересный момент: есть одноэтажный дом с 33 спальнями, что выглядит как аномалия или ошибочное значение.

В целом, основываясь на этих данных, можно сказать, что наиболее типичные дома в этом наборе данных
— это одно- или двухэтажные дома с 3-4 спальнями.

Задача 3.5
Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома.

In [23]:
data.pivot_table(index='condition', columns='grade', values='price', aggfunc='median')


grade            1         3         4         5         6         7   \
condition                                                               
1          142000.0       NaN  150000.0  190000.0  255000.0  403500.0   
2               NaN  280000.0  145000.0  180000.0  235000.0  305000.0   
3               NaN   75000.0  205000.0  234475.0  265000.0  357500.0   
4               NaN       NaN  238525.0  229975.0  295000.0  390000.0   
5               NaN  262000.0       NaN  227450.0  285475.0  456000.0   

grade            8          9          10         11         12         13  
condition                                                                   
1          932500.0        NaN        NaN        NaN        NaN        NaN  
2          429000.0   715000.0  1752500.0        NaN        NaN        NaN  
3          485000.0   689000.0   890000.0  1209500.0  1807500.0  2888000.0  
4          571250.0   823500.0  1030000.0  1685000.0  2125000.0  5750000.0  
5          696000.0  1078000.0  1650000.0  2050000.0  1990000.0        NaN

Эта сводная таблица показывает медианную стоимость домов в зависимости от их состояния (condition) и оценки дома (grade).

Давайте проанализируем эту таблицу:

1. Дома с наивысшей оценкой (13) и в отличном состоянии (4) имеют самую высокую медианную стоимость — 5,750,000. 
2. Дома с наименьшей оценкой (1) и в плохом состоянии (1) имеют наименьшую медианную стоимость — 142,000.
3. Для домов с оценкой 1 и состоянием 2, 3, 4, 5 цены отсутствуют, что указывает на отсутствие таких домов в данных или на то, что такие дома не продавались.
4. С увеличением оценки дома его стоимость, как правило, увеличивается, что логично, так как лучший рейтинг указывает на лучшее качество или расположение дома.
5. Внутри одного и того же рейтинга дома в лучшем состоянии обычно стоят дороже.

На основе этих данных можно сделать вывод, чем выше оценка и состояние дома, тем выше его стоимость.